spec/unit/typesafe/config/config_value_spec.rb

```diff
@@ -0,0 +1,503 @@
+require 'spec_helper'
+require 'hocon'
+require 'test_utils'
+
+require 'hocon/impl/config_delayed_merge'
+require 'hocon/impl/config_delayed_merge_object'
+
+
+
+SimpleConfigOrigin = Hocon::Impl::SimpleConfigOrigin
+SimpleConfigObject = Hocon::Impl::SimpleConfigObject
+SimpleConfigList = Hocon::Impl::SimpleConfigList
+SubstitutionExpression = Hocon::Impl::SubstitutionExpression
+ConfigReference = Hocon::Impl::ConfigReference
+ConfigConcatenation = Hocon::Impl::ConfigConcatenation
+ConfigDelayedMerge = Hocon::Impl::ConfigDelayedMerge
+ConfigDelayedMergeObject = Hocon::Impl::ConfigDelayedMergeObject
+ConfigNotResolvedError = Hocon::ConfigError::ConfigNotResolvedError
+AbstractConfigObject = Hocon::Impl::AbstractConfigObject
+
+describe "SimpleConfigOrigin equality" do
+  context "different origins with the same name should be equal" do
+    let(:a) { SimpleConfigOrigin.new_simple("foo") }
+    let(:same_as_a) { SimpleConfigOrigin.new_simple("foo") }
+    let(:b) { SimpleConfigOrigin.new_simple("bar") }
+
+    context "a equals a" do
+      let(:first_object) { a }
+      let(:second_object) { a }
+      include_examples "object_equality"
+    end
+
+    context "a equals same_as_a" do
+      let(:first_object) { a }
+      let(:second_object) { same_as_a }
+      include_examples "object_equality"
+    end
+
+    context "a does not equal b" do
+      let(:first_object) { a }
+      let(:second_object) { b }
+      include_examples "object_inequality"
+    end
+  end
+end
+
+describe "ConfigInt equality" do
+  context "different ConfigInts with the same value should be equal" do
+    a = TestUtils.int_value(42)
+    same_as_a = TestUtils.int_value(42)
+    b = TestUtils.int_value(43)
+
+    context "a equals a" do
+      let(:first_object) { a }
+      let(:second_object) { a }
+      include_examples "object_equality"
+    end
+
+    context "a equals same_as_a" do
+      let(:first_object) { a }
+      let(:second_object) { same_as_a }
+      include_examples "object_equality"
+    end
+
+    context "a does not equal b" do
+      let(:first_object) { a }
+      let(:second_object) { b }
+      include_examples "object_inequality"
+    end
+  end
+end
+
+describe "ConfigFloat equality" do
+  context "different ConfigInts with the same value should be equal" do
+    a = TestUtils.float_value(42)
```

Comment:
I think these values and test strings need to be changed to be floats instead of ints